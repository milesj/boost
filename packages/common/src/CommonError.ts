import { createScopedError } from '@boost/internal';

const errors = {
  PARSE_INVALID_EXT: 'Unable to parse file "%s". Unsupported file extension.',
  PATH_REQUIRE_ABSOLUTE: 'An absolute file path is required.',
  PATH_RESOLVE_LOOKUPS: 'Failed to resolve a path using the following lookups (in order):\n%s\n',
  PROJECT_NO_PACKAGE: 'No `package.json` found within project root.',
};

export type CommonErrorCode = keyof typeof errors;

export default createScopedError<CommonErrorCode>('CMN', 'CommonError', errors);
