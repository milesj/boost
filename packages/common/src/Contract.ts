import optimal, { Blueprint, Predicates, predicates } from 'optimal';
import { Optionable } from './types';

export default abstract class Contract<T extends object = {}> implements Optionable<T> {
  readonly options: Readonly<Required<T>>;

  constructor(options?: T) {
    this.options = this.configure(options);
  }

  /**
   * Set an options object by merging the new partial and existing options
   * with the defined blueprint, while running all validation checks.
   * Freeze and return the options object.
   */
  configure(options?: Partial<T> | ((options: Required<T>) => Partial<T>)): Readonly<Required<T>> {
    const nextOptions = typeof options === 'function' ? options(this.options) : options;

    // We don't want the options property to be modified directly,
    // so it's read only, but we still want to modify it with this function.
    // @ts-expect-error
    this.options = Object.freeze(
      optimal(
        { ...this.options, ...nextOptions },
        this.blueprint(predicates, this.options === undefined) as Blueprint<T>,
        {
          name: this.constructor.name,
        },
      ),
    );

    return this.options;
  }

  /**
   * Define an optimal blueprint in which to validate and build the
   * options object passed to the constructor, or when manual setting.
   */
  abstract blueprint(predicates: Predicates, onConstruction?: boolean): Blueprint<object>;
}
