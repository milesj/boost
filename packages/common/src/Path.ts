import fs from 'fs';
import path from 'path';
// import { fileURLToPath } from 'url';
import { FilePath, Pathable, PortablePath } from './types';

/**
 * An immutable class for operating on file system paths.
 */
export class Path implements Pathable {
	static DELIMITER = path.delimiter;

	static SEP = path.sep;

	private internalPath: string = '';

	private isNormalized: boolean = false;

	constructor(...parts: PortablePath[]) {
		this.internalPath = path.join(...parts.map(String));
		// this.internalPath = path.join(
		// 	...parts.map((part) => {
		// 		const p = String(part);

		// 		return p.startsWith('file:') ? fileURLToPath(p) : p;
		// 	}),
		// );
	}

	/**
	 * Create and return a new `Path` instance.
	 */
	static create(filePath: PortablePath): Path {
		return new Path(filePath);
	}

	/**
	 * Like `create()` but also resolves the path against a working directory.
	 */
	static resolve(filePath: PortablePath, cwd?: PortablePath): Path {
		return Path.create(filePath).resolve(cwd);
	}

	/**
	 * Append path parts to the end of the current path
	 * and return a new `Path` instance.
	 */
	append(...parts: PortablePath[]): Path {
		return new Path(this.internalPath, ...parts);
	}

	/**
	 * Returns true if both paths are equal using strict equality.
	 */
	equals(filePath: PortablePath): boolean {
		return this.path() === Path.create(filePath).path();
	}

	/**
	 * Return the extension (if applicable) with or without leading period.
	 */
	ext(withoutPeriod: boolean = false): string {
		const ext = path.extname(this.path());

		return withoutPeriod && ext.startsWith('.') ? ext.slice(1) : ext;
	}

	/**
	 * Return true if the current path exists.
	 */
	exists(): boolean {
		return fs.existsSync(this.path());
	}

	/**
	 * Return true if the current path is absolute.
	 */
	isAbsolute(): boolean {
		return path.isAbsolute(this.path());
	}

	/**
	 * Return true if the current path is a folder.
	 */
	isDirectory(): boolean {
		return fs.statSync(this.path()).isDirectory();
	}

	/**
	 * Return true if the current path is a file.
	 */
	isFile(): boolean {
		return fs.statSync(this.path()).isFile();
	}

	/**
	 * Return the file name (with optional extension) or folder name.
	 */
	name(withoutExtension: boolean = false): string {
		let name = path.basename(this.path());

		if (withoutExtension) {
			name = name.replace(this.ext(), '');
		}

		return name;
	}

	/**
	 * Return the parent folder as a new `Path` instance.
	 */
	parent(): Path {
		return new Path(path.dirname(this.internalPath));
	}

	/**
	 * Return the current path as a normalized string.
	 */
	path(): FilePath {
		if (!this.isNormalized) {
			this.isNormalized = true;
			this.internalPath = path.normalize(this.internalPath);
		}

		return this.internalPath;
	}

	/**
	 * Prepend path parts to the beginning of the current path
	 * and return a new `Path` instance.
	 */
	prepend(...parts: PortablePath[]): Path {
		return new Path(...parts, this.internalPath);
	}

	/**
	 * Returns a canonical path by resolving directories and symlinks.
	 */
	// istanbul ignore next
	realPath(): FilePath {
		return fs.realpathSync.native(this.path());
	}

	/**
	 * Return a new relative `Path` instance from the current
	 * "from" path to the defined "to" path.
	 */
	relativeTo(to: PortablePath): Path {
		return new Path(path.relative(this.internalPath, String(to)));
	}

	/**
	 * Return a new `Path` instance where the current path is accurately
	 * resolved against the defined working directory.
	 */
	resolve(cwd?: PortablePath): Path {
		return new Path(path.resolve(String(cwd ?? process.cwd()), this.internalPath));
	}

	toJSON(): FilePath {
		return this.path();
	}

	toString(): FilePath {
		return this.path();
	}
}
