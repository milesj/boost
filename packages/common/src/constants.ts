// https://github.com/npm/validate-npm-package-name
export const MODULE_NAME_PART = /[a-z0-9]{1}[-a-z0-9_.]*/u;

export const MODULE_NAME_PATTERN = new RegExp(
	`^(?:@(${MODULE_NAME_PART.source})/)?(${MODULE_NAME_PART.source})$`,
	'u',
);
