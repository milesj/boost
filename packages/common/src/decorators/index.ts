export { default as Bind } from './Bind';
export { default as Debounce } from './Debounce';
export { default as Deprecate } from './Deprecate';
