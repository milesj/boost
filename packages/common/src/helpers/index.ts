export * from './createBlueprint';
export * from './deepFreeze';
export * from './deepMerge';
export * from './formatMs';
export * from './instanceOf';
export * from './isEmpty';
export * from './isFilePath';
export * from './isModuleName';
export * from './isObject';
export * from './isPlainObject';
export * from './toArray';
