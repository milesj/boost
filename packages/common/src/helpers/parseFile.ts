import fs from 'fs';
import CommonError from '../CommonError';
import Path from '../Path';
import { parse as parseJSON } from '../serializers/json';
import { parse as parseYAML } from '../serializers/yaml';
import { PortablePath } from '../types';
import requireModule from './requireModule';

export default function parseFile<T>(filePath: PortablePath): T {
  const path = Path.create(filePath);

  if (!path.isAbsolute()) {
    throw new CommonError('PATH_REQUIRE_ABSOLUTE');
  }

  switch (path.ext()) {
    case '.js':
    case '.jsx':
      return requireModule(path);

    case '.json':
    case '.json5':
      return parseJSON(fs.readFileSync(path.path(), 'utf8'));

    case '.yml':
    case '.yaml':
      return parseYAML(fs.readFileSync(path.path(), 'utf8'));

    default:
      throw new CommonError('PARSE_INVALID_EXT', [path.name()]);
  }
}
