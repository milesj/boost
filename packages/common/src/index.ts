/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal, { predicates, Blueprint, Predicates } from 'optimal';
import CommonError, { CommonErrorCode } from './CommonError';
import Contract from './Contract';
import ExitError from './ExitError';
import Path from './Path';
import PathResolver from './PathResolver';
import Project from './Project';
import * as json from './serializers/json';
import * as yaml from './serializers/yaml';

export * from './constants';
export * from './helpers';
export * from './types';

export {
  CommonError,
  CommonErrorCode,
  Contract,
  Blueprint,
  ExitError,
  Path,
  PathResolver,
  Predicates,
  Project,
  optimal,
  predicates,
  json,
  yaml,
};
