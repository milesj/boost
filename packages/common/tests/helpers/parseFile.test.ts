import { getFixturePath } from '@boost/test-utils';
import { parseFile } from '../../src/helpers/parseFile';
import { Path } from '../../src/Path';

describe('parseFile()', () => {
	it('errors if path is not absolute', () => {
		expect(() => {
			parseFile('../some/relative/path.js');
		}).toThrowErrorMatchingSnapshot();
	});

	it('errors for unsupported extension', () => {
		expect(() => {
			parseFile(getFixturePath('file-types', 'unknown.ext'));
		}).toThrowErrorMatchingSnapshot();
	});

	it('supports .js extension', () => {
		expect(parseFile(getFixturePath('file-types', 'js.js'))).toEqual({ type: 'js' });
	});

	it('supports .js extension via the `Path` class', () => {
		expect(parseFile(new Path(getFixturePath('file-types', 'js.js')))).toEqual({ type: 'js' });
	});

	it('supports .jsx extension', () => {
		expect(parseFile(getFixturePath('file-types', 'jsx.jsx'))).toEqual({ jsx: true, type: 'js' });
	});

	it('supports .ts extension', () => {
		expect(parseFile(getFixturePath('file-types', 'ts.ts'))).toEqual({ type: 'ts' });
	});

	it('supports .tsx extension', () => {
		expect(parseFile(getFixturePath('file-types', 'tsx.tsx'))).toEqual({ type: 'tsx' });
	});

	it('supports .json extension', () => {
		expect(parseFile(getFixturePath('file-types', 'json.json'))).toEqual({ type: 'json' });
	});

	it('supports .json5 extension', () => {
		expect(parseFile(getFixturePath('file-types', 'json5.json5'))).toEqual({ type: 'json5' });
	});

	it('supports .yaml extension', () => {
		expect(parseFile(getFixturePath('file-types', 'yaml.yaml'))).toEqual({ type: 'yaml' });
	});

	it('supports .yml extension', () => {
		expect(parseFile(getFixturePath('file-types', 'yml.yml'))).toEqual({
			short: true,
			type: 'yaml',
		});
	});
});
