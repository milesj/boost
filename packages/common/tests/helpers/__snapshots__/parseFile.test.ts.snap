// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFile() errors for unsupported extension 1`] = `"Unable to parse file \\"unknown.ext\\". Unsupported file extension. [CMN:PARSE_INVALID_EXT]"`;

exports[`parseFile() errors if path is not absolute 1`] = `"An absolute file path is required. [CMN:PATH_REQUIRE_ABSOLUTE]"`;
